### PR TITLE
feat(dui2/rhino/revit): better reporting and previews

### LINKED ISSUE #1447: feat(dui2): add selectable reporting
Update the current DUI2 reporting so that:

- `ApplicationObject` class contains props to store conversion state and created ids
- Report is moved to a new tab next to `Activity` and `Comments` 
- Report shows a list of `ApplicationObject`s that contain a select button to select in model space
- Fallback to old reporting if connectors are not updated with `ApplicationObject` on send or receive

### LINKED ISSUE #1448: feat(rhino): add previews and selectable reports
- Preview send operations by selecting objects in model space that will be sent
- Preview receive operations by generating preview geometry
- Implement new connector bindings for preview and selecting/deselecting doc objects
- Update send and receive reports with `ApplicationObject`

### LINKED ISSUE #1449: feat(revit): add selectable report
Update send and receive reports with `ApplicationObject`

### PR DESCRIPTION
**Pull request by clairekuang**
## Description

This is a major PR that includes changes in:

**Core** 
- new `ApplicationObject`: this is scoped in [notion](https://www.notion.so/speckle/ApplicationObject-6cfbb143a22844378f3fbd4047bc8f67) and replaces the old `ApplicationPlaceholderObject` class. **Breaking change** Old files with saved streams will not deserialize correctly and would have to be re-saved

**DUI2**
- moved `Report` to a tab in `StreamEditView`: this will display old report logging if no `ApplicationObjects` are saved in `ProgressViewModel` or display a list of `ApplicationObjects` if they are present
- new connector bindings: `PreviewSend`, `PreviewReceive`, `ResetDocument`, and an updated `SelectClientObjects`
- [x] Todo: add copy report button that summarizes all `ApplicationObjects`
- [x] Todo: preview button linked to send/receive button enabled state;
- [ ] Todo: add operation errors to top of report
- [ ] Todo: add progress bar on preview receive

**Rhino**
- Refactored Receive method quite heavily, including the `FlattenCommitObject` method to work with the new `ApplicationObject` for reporting
- Implemented `PreviewSend` and `PreviewReceive`
Known issues: 
- [x] previewing blocks and views, which are created during conversion. Should be resolved with new block helper method and skipping views.

**Revit**
- **IMPORTANT** Refactored conversions to only return a single `ApplicationObject` instead of a List - the old nested elements should be stored in the `converted` and `createdIds` prop of the parent `ApplicationObject`
- Updated report for receive
Known issues: 
- [x] update report for send as well
- [x] check structural conversions as some of these didn't quite make sense with how they were handling the old `ApplicationPlaceholderObject`

**ETABS/CSI/Tekla Structures**
Known Issues:
- [x] report not scrollable
- [x] preview recieve crashes plugin 

Other sharp connectors
- added new connector bindings

> Next steps include potentially refactoring the `ReportObjects` property on `ProgressReport` (which is used to pass `ApplicationObjects` between the connector and converter) and the `ContextObjects` connector property, as there seems to be some overlap of functionality.
Known bugs:
clicking preview receive on and off quickly losing sync state
switching preview receive on and off wipes report (this should be preserved)

- Fixes #1447
- Fixes #1448
- Fixes #1449

## Type of change

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?

- Manual Tests: Rhino sends and receives
- Requires more thorough testing to check for build errors and breaking behavior in:
- [x] Revit @teocomi 
- [x] Structural connectors @Reynold-Chan 
- [x] Bentley

## Docs

- Will be updated ASAP

### CODE REVIEW

**Comment by Reynold-Chan** [comment]
report not scrollable for ETABS

**Comment by Reynold-Chan** [comment]
clicking preview for recieve in CSI causes connector to crash

**Comment by Reynold-Chan** [comment]
looks good with the minor bugs

File: DesktopUI2/DesktopUI2/ViewModels/StreamViewModel.cs [hunk 1]
```diff
@@ -232,6 +227,55 @@ public List<ActivityViewModel> Activity
...
+      get => _report;
+      private set
+      {
+        this.RaiseAndSetIfChanged(ref _report, value);
+        this.RaisePropertyChanged("FilteredReport");
+        this.RaisePropertyChanged("HasReportItems");
+        this.RaisePropertyChanged("Log");
+      }
+    }
+    public List<ApplicationObjectViewModel> FilteredReport
```
Comment (teocomi) [inline DesktopUI2/DesktopUI2/ViewModels/StreamViewModel.cs:242]: TODO later maybe, use a `SourceCache` for filtering instead as it will be more performant on large collections, see: https://github.com/specklesystems/speckle-manager2/blob/main/Speckle.Manager/ViewModels/ConnectorsViewModel.cs#L106

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertWall.cs [hunk 2]
```diff
@@ -115,40 +117,22 @@ public List<ApplicationPlaceholderObject> WallToNative(BuiltElements.Wall speckl
...
-        ApplicationGeneratedId = revitWall.UniqueId,
-        NativeObject = revitWall
-        }
-      };
-
-      var hostedElements = SetHostedElements(speckleWall, revitWall);
-      placeholders.AddRange(hostedElements);
-
-
-      Report.Log($"{(isUpdate ? "Updated" : "Created")} Wall {revitWall.Id}");
```
Comment (teocomi) [inline Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertWall.cs:138]: Any specific reason why was this removed? Didn't check other conversions, but if possible we should keep the old Report in place....

**Review by teocomi — CHANGES_REQUESTED** [review]
In addition to my inline comments, some changes that might be needed:
- preview icon should be disabled when sending/receiving (crashes Revit if clicked)
- preview icon should be disabled if there is nothing to preview (crashes Revit if clicked)
- MAYBE the report tab could be hidden if empty?

**Comment by teocomi** [comment]
I fixed the old report not being scrollable, it was hidden behind the new one! ;)

File: Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertWall.cs (hunk 2 above)
Comment (clairekuang) [reply Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertWall.cs:138]: The new `ApplicationObject` status indicates if the element is updated or created! Unless you don't want me to include the new report for the revit connector?

**Review by Reynold-Chan — APPROVED** [review]
scroll working now :)

**Review by teocomi — APPROVED** [review]
Just tested in Revit this time, looks good!
Some minor improvements for later:
- automatically switch to report tab when clicking on Preview
- add some info text at the top of the report page (eg what it it and how it works) + an empty state
- do not block the UI when generating the report
- hide the preview button if not implemented
- improve performance and readability of large reports (eg revit's)
